pkg/image/api/helper.go

```diff
@@ -236,45 +280,104 @@ func NormalizeImageStreamTag(name string) string {
+			// should never have an empty history, but just in case...
+			return image, nil
+		}
 
-	v1Metadata := DockerV1CompatibilityImage{}
-	if err := json.Unmarshal([]byte(manifest.History[0].DockerV1Compatibility), &v1Metadata); err != nil {
-		return nil, err
-	}
+		v1Metadata := DockerV1CompatibilityImage{}
+		if err := json.Unmarshal([]byte(manifest.History[0].DockerV1Compatibility), &v1Metadata); err != nil {
+			return nil, err
+		}
+
+		image.DockerImageLayers = make([]ImageLayer, len(manifest.FSLayers))
+		for i, layer := range manifest.FSLayers {
+			image.DockerImageLayers[i].Name = layer.DockerBlobSum
+		}
+		if len(manifest.History) == len(image.DockerImageLayers) {
+			image.DockerImageLayers[0].Size = v1Metadata.Size
+			var size = DockerV1CompatibilityImageSize{}
+			for i, obj := range manifest.History[1:] {
+				size.Size = 0
+				if err := json.Unmarshal([]byte(obj.DockerV1Compatibility), &size); err != nil {
+					continue
+				}
+				image.DockerImageLayers[i+1].Size = size.Size
+			}
+		} else {
+			glog.V(4).Infof("Imported image has mismatched layer count and history count, not updating image metadata: %s", image.Name)
+		}
+		// reverse order of the layers for v1 (lowest = 0, highest = i)
+		for i, j := 0, len(image.DockerImageLayers)-1; i < j; i, j = i+1, j-1 {
+			image.DockerImageLayers[i], image.DockerImageLayers[j] = image.DockerImageLayers[j], image.DockerImageLayers[i]
+		}
 
-	image.DockerImageMetadata.ID = v1Metadata.ID
-	image.DockerImageMetadata.Parent = v1Metadata.Parent
-	image.DockerImageMetadata.Comment = v1Metadata.Comment
-	image.DockerImageMetadata.Created = v1Metadata.Created
-	image.DockerImageMetadata.Container = v1Metadata.Container
-	image.DockerImageMetadata.ContainerConfig = v1Metadata.ContainerConfig
-	image.DockerImageMetadata.DockerVersion = v1Metadata.DockerVersion
-	image.DockerImageMetadata.Author = v1Metadata.Author
-	image.DockerImageMetadata.Config = v1Metadata.Config
-	image.DockerImageMetadata.Architecture = v1Metadata.Architecture
-	image.DockerImageMetadata.Size = v1Metadata.Size
+		image.DockerImageMetadata.ID = v1Metadata.ID
+		image.DockerImageMetadata.Parent = v1Metadata.Parent
+		image.DockerImageMetadata.Comment = v1Metadata.Comment
+		image.DockerImageMetadata.Created = v1Metadata.Created
+		image.DockerImageMetadata.Container = v1Metadata.Container
+		image.DockerImageMetadata.ContainerConfig = v1Metadata.ContainerConfig
+		image.DockerImageMetadata.DockerVersion = v1Metadata.DockerVersion
+		image.DockerImageMetadata.Author = v1Metadata.Author
+		image.DockerImageMetadata.Config = v1Metadata.Config
+		image.DockerImageMetadata.Architecture = v1Metadata.Architecture
+		if len(image.DockerImageLayers) > 0 {
+			size := int64(0)
+			for _, layer := range image.DockerImageLayers {
+				size += layer.Size
+			}
+			image.DockerImageMetadata.Size = size
+		} else {
+			image.DockerImageMetadata.Size = v1Metadata.Size
+		}
+	case 2:
+		// TODO: need to prepare for this
```

Comment:
we have to identify the content type of the manifest from config, then parse and find the reference manifests.  We'll probably want to coalesce multiple manifests into a single image (for our purposes) rather than delegating to the blob store, but needs more thought.  We'll still need to extract effective image data to the client.